Fix build warnings and remove dead code

defuse opened the pull request:
Closes #911. Closes #926. Closes #457.

**WARNING: I force pushed**

defuse commented:
I wonder if any of Sean's not-yet-posted PRs will use any of the util.cpp functions I'm removing. If so, then I'll add them back one-by-one making sure they're implemented properly.

defuse commented:
For review.

str4d commented on src/crypto/equihash.cpp:
```diff
@@ -247,7 +246,7 @@ std::set<std::vector<eh_index>> Equihash::BasicSolve(const eh_HashState& base_st
 bool Equihash::IsValidSolution(const eh_HashState& base_state, std::vector<eh_index> soln)
 {
-    eh_index soln_size { pow(2, k) };
+    eh_index soln_size { 1u << k };
```
I already make this change in my optimised Equihash PR.

str4d replied:
(although I don't make it explicitly unsigned - is there any benefit?)

defuse replied:
Yes, there's a compiler warning if it's not explicitly unsigned (probably for the case where k=31)

ebfull commented on src/test/sighash_tests.cpp:
```diff
@@ -130,8 +130,8 @@ void static RandomTransaction(CMutableTransaction &tx, bool fSingle) {
             pourtx.randomSeed = GetRandHash();
-            pourtx.ciphertexts[0] = {insecure_rand() % 100, insecure_rand() % 100};
-            pourtx.ciphertexts[1] = {insecure_rand() % 100, insecure_rand() % 100};
+            pourtx.ciphertexts[0] = {(unsigned char)insecure_rand(), (unsigned char)insecure_rand()};
```
I think this should just use libsodium's randombytes.

ebfull commented:
ACK once the entire ciphertext is filled with random bytes using libsodium.

daira commented:
ACK.

defuse commented:
Okay, I used randombytes_buf to fill the whole buffer.

Arch Linux upgraded me to GCC 6, which broke the build. I fixed it with a patch to libsnark (upstream PR: https://github.com/scipr-lab/libsnark/pull/35) in the depends system. It also added a new type of warning that the boost headers generate (they are some of the most horribly convoluted warnings I've ever seen in my life), so I added a flag to ignore those.

defuse commented:
In case you're wondering, here's what it looks like:

```
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:318:13: error: placement new constructing an object of type ‘boost::detail::function::functor_manager_common<boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, char>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > > >::functor_type {aka boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, char>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >}’ and size ‘24’ in a region of type ‘char’ and size ‘1’ [-Werror=placement-new=]
             new (reinterpret_cast<void*>(&out_buffer.data)) functor_type(*in_functor);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp: In instantiation of ‘static void boost::detail::function::functor_manager_common<Functor>::manage_small(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >]’:
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:364:56:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, mpl_::true_) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >; mpl_::true_ = mpl_::bool_<true>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:412:18:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, boost::detail::function::function_obj_tag) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:440:20:   required from ‘static void boost::detail::function::functor_manager<Functor>::manage(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:934:13:   required from ‘void boost::function2<R, T1, T2>::assign_to(Functor) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >; R = void; T0 = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; T1 = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:722:7:   required from ‘boost::function2<R, T1, T2>::function2(Functor, typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >; R = void; T0 = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; T1 = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type = int]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:1071:16:   [ skipping 8 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:133:48:   required from ‘static boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::impl::phrase_parser<boost::spirit::classic::space_parser>::parse(const IteratorT&, const IteratorT&, const ParserT&, const boost::spirit::classic::space_parser&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:155:18:   required from ‘boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::parse(const IteratorT&, const IteratorT&, const boost::spirit::classic::parser<DerivedT>&, const boost::spirit::classic::parser<SkipT>&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>; SkipT = boost::spirit::classic::space_parser]’
./json/json_spirit_reader_template.h:523:58:   required from ‘Iter_type json_spirit::read_range_or_throw(Iter_type, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:551:40:   required from ‘bool json_spirit::read_range(Iter_type&, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:572:34:   required from ‘bool json_spirit::read_string(const String_type&, Value_type&) [with String_type = std::__cxx11::basic_string<char>; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
gtest/test_merkletree.cpp:35:33:   required from here
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:318:13: error: placement new constructing an object of type ‘boost::detail::function::functor_manager_common<boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > > >::functor_type {aka boost::_bi::bind_t<void, boost::_mfi::mf2<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::_bi::list3<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1>, boost::arg<2> > >}’ and size ‘24’ in a region of type ‘char’ and size ‘1’ [-Werror=placement-new=]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp: In instantiation of ‘static void boost::detail::function::functor_manager_common<Functor>::manage_small(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’:
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:364:56:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, mpl_::true_) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; mpl_::true_ = mpl_::bool_<true>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:412:18:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, boost::detail::function::function_obj_tag) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:440:20:   required from ‘static void boost::detail::function::functor_manager<Functor>::manage(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:934:13:   required from ‘void boost::function1<R, T1>::assign_to(Functor) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; R = void; T0 = double]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:722:7:   required from ‘boost::function1<R, T1>::function1(Functor, typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; R = void; T0 = double; typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type = int]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:1071:16:   [ skipping 8 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:133:48:   required from ‘static boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::impl::phrase_parser<boost::spirit::classic::space_parser>::parse(const IteratorT&, const IteratorT&, const ParserT&, const boost::spirit::classic::space_parser&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:155:18:   required from ‘boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::parse(const IteratorT&, const IteratorT&, const boost::spirit::classic::parser<DerivedT>&, const boost::spirit::classic::parser<SkipT>&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>; SkipT = boost::spirit::classic::space_parser]’
./json/json_spirit_reader_template.h:523:58:   required from ‘Iter_type json_spirit::read_range_or_throw(Iter_type, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:551:40:   required from ‘bool json_spirit::read_range(Iter_type&, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:572:34:   required from ‘bool json_spirit::read_string(const String_type&, Value_type&) [with String_type = std::__cxx11::basic_string<char>; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
gtest/test_merkletree.cpp:35:33:   required from here
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:318:13: error: placement new constructing an object of type ‘boost::detail::function::functor_manager_common<boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > > >::functor_type {aka boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, double>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >}’ and size ‘24’ in a region of type ‘char’ and size ‘1’ [-Werror=placement-new=]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp: In instantiation of ‘static void boost::detail::function::functor_manager_common<Functor>::manage_small(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’:
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:364:56:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, mpl_::true_) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; mpl_::true_ = mpl_::bool_<true>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:412:18:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, boost::detail::function::function_obj_tag) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:440:20:   required from ‘static void boost::detail::function::functor_manager<Functor>::manage(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:934:13:   required from ‘void boost::function1<R, T1>::assign_to(Functor) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; R = void; T0 = long int]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:722:7:   required from ‘boost::function1<R, T1>::function1(Functor, typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; R = void; T0 = long int; typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type = int]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:1071:16:   [ skipping 8 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:133:48:   required from ‘static boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::impl::phrase_parser<boost::spirit::classic::space_parser>::parse(const IteratorT&, const IteratorT&, const ParserT&, const boost::spirit::classic::space_parser&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:155:18:   required from ‘boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::parse(const IteratorT&, const IteratorT&, const boost::spirit::classic::parser<DerivedT>&, const boost::spirit::classic::parser<SkipT>&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>; SkipT = boost::spirit::classic::space_parser]’
./json/json_spirit_reader_template.h:523:58:   required from ‘Iter_type json_spirit::read_range_or_throw(Iter_type, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:551:40:   required from ‘bool json_spirit::read_range(Iter_type&, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:572:34:   required from ‘bool json_spirit::read_string(const String_type&, Value_type&) [with String_type = std::__cxx11::basic_string<char>; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
gtest/test_merkletree.cpp:35:33:   required from here
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:318:13: error: placement new constructing an object of type ‘boost::detail::function::functor_manager_common<boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > > >::functor_type {aka boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >}’ and size ‘24’ in a region of type ‘char’ and size ‘1’ [-Werror=placement-new=]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp: In instantiation of ‘static void boost::detail::function::functor_manager_common<Functor>::manage_small(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’:
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:364:56:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, mpl_::true_) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; mpl_::true_ = mpl_::bool_<true>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:412:18:   required from ‘static void boost::detail::function::functor_manager<Functor>::manager(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type, boost::detail::function::function_obj_tag) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:440:20:   required from ‘static void boost::detail::function::functor_manager<Functor>::manage(const boost::detail::function::function_buffer&, boost::detail::function::function_buffer&, boost::detail::function::functor_manager_operation_type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:934:13:   required from ‘void boost::function1<R, T1>::assign_to(Functor) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; R = void; T0 = long unsigned int]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:722:7:   required from ‘boost::function1<R, T1>::function1(Functor, typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type) [with Functor = boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >; R = void; T0 = long unsigned int; typename boost::enable_if_c<boost::type_traits::ice_not<boost::is_integral<Functor>::value>::value, int>::type = int]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_template.hpp:1071:16:   [ skipping 8 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:133:48:   required from ‘static boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::impl::phrase_parser<boost::spirit::classic::space_parser>::parse(const IteratorT&, const IteratorT&, const ParserT&, const boost::spirit::classic::space_parser&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>]’
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/spirit/home/classic/core/scanner/impl/skipper.ipp:155:18:   required from ‘boost::spirit::classic::parse_info<IteratorT> boost::spirit::classic::parse(const IteratorT&, const IteratorT&, const boost::spirit::classic::parser<DerivedT>&, const boost::spirit::classic::parser<SkipT>&) [with IteratorT = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; ParserT = boost::spirit::classic::sequence<json_spirit::Json_grammer<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, boost::spirit::classic::end_parser>; SkipT = boost::spirit::classic::space_parser]’
./json/json_spirit_reader_template.h:523:58:   required from ‘Iter_type json_spirit::read_range_or_throw(Iter_type, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:551:40:   required from ‘bool json_spirit::read_range(Iter_type&, Iter_type, Value_type&) [with Iter_type = __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> >; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
./json/json_spirit_reader_template.h:572:34:   required from ‘bool json_spirit::read_string(const String_type&, Value_type&) [with String_type = std::__cxx11::basic_string<char>; Value_type = json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >]’
gtest/test_merkletree.cpp:35:33:   required from here
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/function/function_base.hpp:318:13: error: placement new constructing an object of type ‘boost::detail::function::functor_manager_common<boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > > >::functor_type {aka boost::_bi::bind_t<void, boost::_mfi::mf1<void, json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >, long unsigned int>, boost::_bi::list2<boost::_bi::value<json_spirit::Semantic_actions<json_spirit::Value_impl<json_spirit::Config_vector<std::__cxx11::basic_string<char> > >, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > >*>, boost::arg<1> > >}’ and size ‘24’ in a region of type ‘char’ and size ‘1’ [-Werror=placement-new=]
cc1plus: error: unrecognized command line option ‘-Wno-replacement-new’ [-Werror]
cc1plus: all warnings being treated as errors
```

defuse commented:
NACK because the scope of those `-Wno-*` flags should really be limited to Boost (or I need to figure out how to downgrade my compiler or something).

defuse commented:
@zkbot try

defuse commented:
I retract my self-NACK. I think we should just do this.

daira commented:
ACK https://github.com/zcash/zcash/pull/939/commits/be1bae14e8a769966b833026ed47a43f5250a4b9

nathan-at-least commented on zcutil/build.sh:
```diff
@@ -35,5 +35,5 @@ PREFIX="$(pwd)/depends/x86_64-unknown-linux-gnu/"
 make "$@" -C ./depends/ V=1 NO_QT=1
 ./autogen.sh
-./configure --prefix="${PREFIX}" --with-gui=no "$LCOV_ARG" CXXFLAGS='-O0 -g'
+./configure --prefix="${PREFIX}" --with-gui=no "$LCOV_ARG" CXXFLAGS='-Og -g'
```
Note: @defuse tells me this flag has sufficient optimization for the source fortify stuff, while also allowing debugging.

nathan-at-least commented on src/Makefile.am:
```diff
@@ -415,7 +415,7 @@ libzerocash_a_SOURCES = \
   zcash/util.cpp
 
-libzerocash_a_CPPFLAGS = -fPIC -DCURVE_ALT_BN128 -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -std=c++11 -pipe -O2 -O0 -g -Wstack-protector -fstack-protector-all -fPIE -fvisibility=hidden -DSTATIC $(BITCOIN_INCLUDES)
+libzerocash_a_CPPFLAGS = -fPIC -DCURVE_ALT_BN128 -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS $(HARDENED_CPPFLAGS) -std=c++11 -pipe -O2 -O0 -g -Wstack-protector -fstack-protector-all -fPIE -fvisibility=hidden -DSTATIC $(BITCOIN_INCLUDES)
```
Note: Makefile bugfix.

nathan-at-least commented on zcutil/build.sh:
```diff
@@ -35,5 +35,5 @@ PREFIX="$(pwd)/depends/x86_64-unknown-linux-gnu/"
 make "$@" -C ./depends/ V=1 NO_QT=1
 ./autogen.sh
-./configure --prefix="${PREFIX}" --with-gui=no "$LCOV_ARG" CXXFLAGS='-Wno-deprecated-declarations -Werror -Og -g'
+./configure --prefix="${PREFIX}" --with-gui=no "$LCOV_ARG" CXXFLAGS='-Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -Og -g'
```
Note: here's a [reference on ``-Wno-placement-new``](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html) (HT @defuse)

nathan-at-least replied:
This concerns me because these kinds of placement new warnings can reveal potential memory corruption bugs. Meanwhile, boost is full of these warnings, so that's quite unfortunate.

I accept this for now. If we could disable this warning for only boost includes, then we could verify our own software much better. Yet, we'd still be open to memory corruption bugs in boost losing user funds, so ideally boost would be patched to fix such warnings.

nathan-at-least replied:
This [gcc page](https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html) mentions ``-Wno-terminate``.

nathan-at-least commented:
ACK on this PR, even though it introduces some potential for missing or introducing memory corruption or other bugs!  The reason it's worth it is that by making many (hopefully false) positives from boost et all quiet we can start to notice and trust and fix the remaining warnings so we might catch more true-positives.

nathan-at-least commented:
@zkbot r+

ebfull commented:
@zkbot retry

ebfull commented:
@zkbot r+

ebfull commented:
@zkbot r-

ebfull commented:
@zkbot r+ be1bae14e8a769966b833026ed47a43f5250a4b9

ebfull commented:
It was in `try` mode when it was `r+`'d. Let's stop using `try` please!

ebfull commented:
@zkbot r+
